### PR TITLE
[core] Retrieve token from server in GCS client [4/n]

### PR DESCRIPTION
**Pull request by vitsai**
Retrieve the token from the GCS server in the GCS client while connecting, to attach to metadata in requests.

Previous PR (GCS server): https://github.com/ray-project/ray/pull/36535
Next PR (auth): https://github.com/ray-project/ray/pull/36073

## Related issue number

#34763 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 1]
```diff
@@ -56,12 +71,16 @@ GcsServer::GcsServer(const ray::gcs::GcsServerConfig &config,
...
-    gcs_table_storage_ = std::make_shared<InMemoryGcsTableStorage>(main_service_);
+  switch(storage_type_) {
+    case StorageType::IN_MEMORY:
+      gcs_table_storage_ = std::make_shared<InMemoryGcsTableStorage>(main_service_);
+      break;
+    case StorageType::REDIS_PERSIST:
+      gcs_table_storage_ = std::make_shared<gcs::RedisGcsTableStorage>(GetOrConnectRedis());
+      break;
+    default:
+      RAY_CHECK(false) << "Unexpected storage type.";
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:82]: `RAY_LOG(FATAL)`

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 2]
```diff
@@ -503,16 +524,13 @@ void GcsServer::InitUsageStatsClient() {
...
 void GcsServer::InitKVManager() {
-  std::unique_ptr<InternalKVInterface> instance;
   // TODO (yic): Use a factory with configs
-  if (storage_type_ == "redis") {
-    instance = std::make_unique<StoreClientInternalKV>(
-        std::make_unique<RedisStoreClient>(GetOrConnectRedis()));
-  } else if (storage_type_ == "memory") {
-    instance =
-        std::make_unique<StoreClientInternalKV>(std::make_unique<ObservableStoreClient>(
+  std::unique_ptr<InternalKVInterface> instance =
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:528]: Maybe change it to switch-case?

File: src/ray/gcs/gcs_server/pubsub_handler.h [hunk 3]
```diff
@@ -51,6 +54,10 @@ class InternalPubSubHandler : public rpc::InternalPubSubHandler {
   void RemoveSubscriberFrom(const std::string &sender_id);
 
  private:
+  /// Whether to handshake client with server token.
+  bool const token_handshake_;
+  /// GCS server token to pass to subscribers.
+  std::string const &gcs_token_;
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/pubsub_handler.h:60]: I think we don't need to put the server token here. It should be in the RPC layer, not application layer I think.

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 4]
```diff
@@ -32,14 +32,29 @@
...
+    default:
+      UNREACHABLE;
+  }
+}
+
 GcsServer::GcsServer(const ray::gcs::GcsServerConfig &config,
                      instrumented_io_context &main_service)
-    : config_(config),
-      storage_type_(StorageType()),
+    : server_token_(GenerateUUIDV4()),
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:55]: We should generate the token later to make restart work.

**Review by fishbone — COMMENTED** [review]
Thanks for cleaning up the code! I think there are two things here:

- The token shouldn't be generated in the first step. We should connect to the store first and then try to fetch the token. If it's not there, we generate and store it.
- The token should be in the RPC layer, not the application layer. It means we should add it to client context in client call. And in grpc server whenever we got a call, we should check the token first and if it's good, we shall proceed by calling the callback otherwise just return with failure.

I think this PR could just focus on the first one.

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 4 above)
Comment (larrylian) [inline src/ray/gcs/gcs_server/gcs_server.cc:55]: "server_token_" needs to print log for easier problem debug in the future.

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 5]
```diff
@@ -121,6 +145,36 @@ void GcsServer::Start() {
   gcs_init_data->AsyncLoad([this, gcs_init_data] { DoStart(*gcs_init_data); });
 }
 
+void GcsServer::CacheAndSetServerToken() {
+  static std::string const kTokenNamespace = "cluster";
```
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_server.cc:149]: huh, do we need to differentiate the in_memory vs persist storage? I thought we can always assume that the cluster is new if the token is missing.

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 6]
```diff
@@ -121,6 +145,36 @@ void GcsServer::Start() {
...
+  static std::string const kTokenKey = "server_uuid";
+  switch (storage_type_) {
+    case StorageType::IN_MEMORY:
+      token_promise_.set_value(GenerateUUIDV4());
```
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_server.cc:153]: how about we introduce class concept cluster_id into Ray?
https://sourcegraph.com/github.com/ray-project/ray/-/blob/src/ray/common/id.h

**Comment by rkooo567** [comment]
+1 in just introducing the cluster id.

**Comment by jjyao** [comment]
+1 for cluster_id concept

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 5 above)
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_server.cc:149]: Yes, one codepath covers both, but checking and directly setting is an optimization. Up to you which one makes more sense.

**Comment by vitsai** [comment]
1. I've changed it to a cluster-id concept. Based on how the change works now, it would be possible for each ClientCallManager to have an ID corresponding just to the set of services it subscribes to (e.g. ObjectManager and GCS have different IDs that their clients use to authenticate with them). The name "cluster-id" implies something stronger, that a single ClusterID should be generated by GCS and propagated to all the services to use as the same token. 

It doesn't matter for now since the change doesn't plumb through the Cluster-ID propagation to other services, but in the long run it would probably be good for each service to have some sort of verification/auth with its client. But this decision would affect whether or not we need a shared_future for ClientCallManager.

2. Right now, any client without an authentication token can request to RegisterClient() from the GCS. It should ideally be a handshake where the client provides some useful information about its identity, but I think that can be a follow-up.

File: src/ray/protobuf/gcs_service.proto [hunk 7]
```diff
@@ -177,6 +177,13 @@ service ActorInfoGcsService {
   rpc KillActorViaGcs(KillActorViaGcsRequest) returns (KillActorViaGcsReply);
 }
 
+message RegisterClientRequest {}
+
+message RegisterClientReply {
+  GcsStatus status = 1;
+  string cluster_id = 2;
```
Comment (iycheng) [inline src/ray/protobuf/gcs_service.proto:184]: considering bytes? I think we store object id, node id as bytes in protobuf.

**Comment by vitsai** [comment]
OK a few notes about knobs (describing upcoming push). They are kind of redundant and could be organized better:

On the ClientCall side: 
- there is the IS_INSECURE argument that controls whether the cluster_id future will get blocked on before being passed to the CallMethod. If the cluster_id future is not valid, there will simply be no auth regardless of IS_INSECURE.

This future is passed in during initial ClientCallManager creation.

On the ServerCall side:
There is an AuthType enum template parameter that controls the type of authorization. There is also a cluster_id pointer but there are checks that its validity is consistent with AuthType. Unfortunately, there is a third knob, cluster_id_future, and a parameter in RegisterService that checks whether to block on that future before setting up the ServerCallFactory.

This has to do with the initialization order of kv-store, cluster ID, then the remaining services. Can maybe also rework with templates so one parameter is just propagated all the way through.

I believe in the current state there are too many knobs at all the different levels of abstraction and too few checks of consistency, making it hard to figure out whether or not auth is enabled. This should be reworked. Ideally we want (cluster_id is present <=> AuthType != NO_AUTH) in the entire call stack.

But also, priority is just to have a working fix.

File: src/ray/rpc/worker/core_worker_server.h [hunk 8]
```diff
@@ -114,7 +111,8 @@ class CoreWorkerGrpcService : public GrpcService {
 
   void InitServerCallFactories(
       const std::unique_ptr<grpc::ServerCompletionQueue> &cq,
-      std::vector<std::unique_ptr<ServerCallFactory>> *server_call_factories) override {
+      std::vector<std::unique_ptr<ServerCallFactory>> *server_call_factories,
+      ClusterID const *const cluster_id) override {
```
Comment (iycheng) [inline src/ray/rpc/worker/core_worker_server.h:115]: nit. But I think we should just pass by value. By default assign cluster_id = ClusterID::Nil

File: src/ray/gcs/gcs_server/gcs_node_manager.h [hunk 9]
```diff
@@ -49,6 +49,11 @@ class GcsNodeManager : public rpc::NodeInfoHandler {
                           std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage,
                           std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool);
 
+  /// Handle register rpc request come from raylet.
+  void HandleRegisterClient(rpc::RegisterClientRequest request,
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_node_manager.h:53]: How about just rename it to GetClusterID ?

**Comment by iycheng** [comment]
Went through the code in high level. Overall it's simple to follow.

One of my question is that, do you think it's worth using future for performance given the client is only initialized once and we have blocking calls in python too (we can also use multi get to get it directly). So the overhead shouldn't be big.

My concern is that ray's threading model is not good and it's very easy to fail to track where the promise is set and this increase the complexity of the code. For example, if it's in the same thread as the the current one, basically there will be a deadlock.

**Comment by vitsai** [comment]
Added GcsAuthTest to ensure Auth succeeds in persisted case and fails in unpersisted case, but succeeds after reconnecting.

**Comment by vitsai** [comment]
Fixed some bugs.

Outstanding:
- Separating out into two PRs
- Dependency on #35530 (change has been included here as well)
- Refactoring ClusterID const *const to ClusterID const& const
- One token generating codepath for both in-memory and persisted

File: src/ray/rpc/server_call.h [hunk 10]
```diff
@@ -161,6 +175,7 @@ class ServerCallImpl : public ServerCall {
         response_writer_(&context_),
         io_service_(io_service),
         call_name_(std::move(call_name)),
+        cluster_id_(cluster_id),
```
Comment (sriram-anyscale) [inline src/ray/rpc/server_call.h:178]: This line has a different semantics when you change from a pointer to a reference. This uses a copy constructor which is defined in id.h and creates a new object.

File: src/ray/rpc/server_call.h [hunk 11]
```diff
@@ -369,19 +415,21 @@ class ServerCallFactoryImpl : public ServerCallFactory {
         cq_(cq),
         io_service_(io_service),
         call_name_(std::move(call_name)),
+        cluster_id_(cluster_id),
```
Comment (sriram-anyscale) [inline src/ray/rpc/server_call.h:431]: This line has a different semantics when you change from a pointer to a reference. This uses a copy constructor which is defined in id.h and creates a new object.

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 12]
```diff
@@ -81,12 +81,29 @@ void GcsSubscriberClient::PubsubCommandBatch(
 GcsClient::GcsClient(const GcsClientOptions &options, UniqueID gcs_client_id)
     : options_(options), gcs_client_id_(gcs_client_id) {}
 
-Status GcsClient::Connect(instrumented_io_context &io_service) {
+Status GcsClient::Connect(instrumented_io_context &io_service,
+                          ClusterID const &cluster_id) {
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.cc:85]: Change  to `const ClusterID&` to follow the convention of the existing code.

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 13]
```diff
@@ -81,12 +81,29 @@ void GcsSubscriberClient::PubsubCommandBatch(
...
+  cluster_token_promise_ = std::promise<ClusterID>();
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.cc:86]: Add comments about why we have to use promise and do thing in async way instead of doing it in sync way.

**Comment by vitsai** [comment]
Split out authentication logic to #36073

File: src/ray/gcs/gcs_client/gcs_client.h [hunk 14]
```diff
@@ -81,7 +86,8 @@ class RAY_EXPORT GcsClient : public std::enable_shared_from_this<GcsClient> {
   /// This function must be called before calling other functions.
   ///
   /// \return Status
-  virtual Status Connect(instrumented_io_context &io_service);
+  virtual Status Connect(instrumented_io_context &io_service,
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/gcs_client.h:89]: Could you document the params?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 15]
```diff
@@ -122,6 +147,31 @@ void GcsServer::Start() {
   gcs_init_data->AsyncLoad([this, gcs_init_data] { DoStart(*gcs_init_data); });
 }
 
+void GcsServer::CacheAndSetClusterId() {
+  static std::string const kTokenNamespace = "cluster";
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server.cc:151]: Can we define both constants (kTokenNamespace, kClusterIdKey) in the same place?

File: src/ray/rpc/client_call.h [hunk 16]
```diff
@@ -67,6 +69,7 @@ class ClientCallImpl : public ClientCall {
   ///
   /// \param[in] callback The callback function to handle the reply.
   explicit ClientCallImpl(const ClientCallback<Reply> &callback,
+                          ClusterID const *cluster_token,
```
Comment (jjyao) [inline src/ray/rpc/client_call.h:72]: std::optional?

File: src/ray/common/status.h [hunk 17]
```diff
@@ -115,8 +115,7 @@ enum class StatusCode : char {
   ObjectUnknownOwner = 29,
   RpcError = 30,
   OutOfResource = 31,
-  // Meaning the ObjectRefStream reaches to the end of stream.
```
Comment (iycheng) [inline src/ray/common/status.h:118]: put it back?

File: src/ray/rpc/client_call.h [hunk 18]
```diff
@@ -309,6 +331,9 @@ class ClientCallManager {
     }
   }
 
+  /// UUID of this generation of the cluster.
```
Comment (jjyao) [inline src/ray/rpc/client_call.h:334]: what does this sentence mean?

File: src/ray/rpc/worker/core_worker_server.h [hunk 19]
```diff
@@ -43,8 +43,6 @@ namespace rpc {
       CoreWorkerService, UpdateObjectLocationBatch, -1)                                  \
   RPC_SERVICE_HANDLER_SERVER_METRICS_DISABLED(                                           \
       CoreWorkerService, GetObjectLocationsOwner, -1)                                    \
-  RPC_SERVICE_HANDLER_SERVER_METRICS_DISABLED(                                           \
-      CoreWorkerService, ReportGeneratorItemReturns, -1)                                 \
```
Comment (iycheng) [inline src/ray/rpc/worker/core_worker_server.h:47]: wrong change?

File: src/ray/rpc/gcs_server/gcs_rpc_server.h [hunk 20]
```diff
@@ -75,6 +75,7 @@ using namespace rpc::autoscaler;
                       HANDLER,                            \
                       RayConfig::instance().gcs_max_active_rpcs_per_handler())
 
+// As this is internal to GCS, there is no need to do cluster auth here.
```
Comment (jjyao) [inline src/ray/rpc/gcs_server/gcs_rpc_server.h:78]: gcs kv store is used by everyone not just gcs itself

File: src/ray/rpc/server_call.h [hunk 21]
```diff
@@ -298,6 +302,10 @@ class ServerCallImpl : public ServerCall {
   /// Human-readable name for this RPC call.
   std::string call_name_;
 
+  /// ID of the cluster to check incoming RPC calls against.
+  /// Check skipped if empty.
```
Comment (jjyao) [inline src/ray/rpc/server_call.h:306]: Should we check cluster for all the rpc calls happening in the cluster except the RegisterClient one?

File: src/ray/rpc/grpc_server.h [hunk 22]
```diff
@@ -77,11 +78,26 @@ class GrpcServer {
...
   ///  will be chosen.
+  ///
+  /// Note: Ideally with C++20 we could use constraints here, or auto&& in the contructor.
+  template <typename T = std::shared_future<ClusterID>,
+            typename = typename std::enable_if_t<
+                std::is_convertible<T, std::shared_future<ClusterID>>::value>>
   GrpcServer(std::string name,
              const uint32_t port,
              bool listen_to_localhost_only,
+             T &&cluster_token_future = T(),
```
Comment (iycheng) [inline src/ray/rpc/grpc_server.h:89]: I feel this is a little bit confuse. Let's just add another constructor for this.

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 23]
```diff
@@ -32,21 +32,37 @@
...
 GcsServer::GcsServer(const ray::gcs::GcsServerConfig &config,
                      instrumented_io_context &main_service)
     : config_(config),
-      storage_type_(StorageType()),
+      storage_type_(GetStorageType()),
       main_service_(main_service),
       rpc_server_(config.grpc_server_name,
                   config.grpc_server_port,
                   config.node_ip_address == "127.0.0.1",
+                  cluster_token_promise_.get_future(),
```
[the remark this answers is not in the document]
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_server.cc:62]: The main thread will be blocked on `future.get()`, but my understanding is that the `promise.set_value()` is executed by grpc on its private executors since it is a reply callback. The alternative to future is to use the sync API, but it is actually the same under the hood: sync API calls async API, and then waits for the future anyway. 

In the deadlocks that I encountered in this change, all of them were related to when `instrumented_io_context.run()` is called (IO event loop started) relative to when the RPC is sent, and not related to using future.

Question of code complexity is valid, though.

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 24]
```diff
@@ -81,12 +81,33 @@ void GcsSubscriberClient::PubsubCommandBatch(
...
   // Connect to gcs service.
-  client_call_manager_ = std::make_unique<rpc::ClientCallManager>(io_service);
+  client_call_manager_ = std::make_unique<rpc::ClientCallManager>(
+      io_service, cluster_token_promise_.get_future());
   gcs_rpc_client_ = std::make_shared<rpc::GcsRpcClient>(
       options_.gcs_address_, options_.gcs_port_, *client_call_manager_);
 
+  // Note: We need to initialize this so we can use the RPC.
+  node_accessor_ = std::make_unique<NodeInfoAccessor>(this);
```
[the remark this answers is not in the document]
Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc:98]: The RegisterClient call goes through `node_accessor_`. I agree the asymmetry with other accessors is weird and frankly undesirable

File: src/ray/rpc/server_call.h (hunk 21 above)
Comment (vitsai) [reply src/ray/rpc/server_call.h:306]: Yes, that's in the followup PR! This one was getting too big, lol

File: src/ray/rpc/client_call.h (hunk 18 above)
Comment (vitsai) [reply src/ray/rpc/client_call.h:334]: If the cluster restarts, it is a new generation of the same cluster... should it just be considered a "new cluster"?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 25]
```diff
@@ -502,26 +549,36 @@ void GcsServer::InitFunctionManager() {
 }
 
 void GcsServer::InitUsageStatsClient() {
-  usage_stats_client_ = std::make_unique<UsageStatsClient>(
-      "127.0.0.1:" + std::to_string(GetPort()), main_service_);
+  // Note: We pass in cluster_id here to avoid deadlock during server init.
```
[the remark this answers is not in the document]
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_server.cc:552]: Yeah, agreed

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 15 above)
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_server.cc:151]: Right now nobody else uses kTokenNamespace, though

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 25 above)
[the remark this answers is not in the document]
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_server.cc:552]: We can do that in a follow-up, though, as auth gets slowly rolled out to all the clients

File: src/ray/rpc/client_call.h [hunk 26]
```diff
@@ -76,6 +79,9 @@ class ClientCallImpl : public ClientCall {
           std::chrono::system_clock::now() + std::chrono::milliseconds(timeout_ms);
       context_.set_deadline(deadline);
     }
+    if (cluster_token) {
```
[the remark this answers is not in the document]
Comment (vitsai) [reply src/ray/rpc/client_call.h:82]: In the initial RegisterClient (GetClusterId), we don't have it yet

File: src/ray/rpc/grpc_server.h (hunk 22 above)
Comment (vitsai) [reply src/ray/rpc/grpc_server.h:89]: Because of the perfect forwarding?

Comment (iycheng) [reply src/ray/rpc/grpc_server.h]: I see. I think it's OK, it's just a ClusterID (28bytes) and only one time thing :)

File: src/ray/rpc/client_call.h (hunk 18 above)
Comment (jjyao) [reply src/ray/rpc/client_call.h:334]: Yea, if it's GCS not backed by redis then after restart, it's a new cluster

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 27]
```diff
@@ -81,12 +81,33 @@ void GcsSubscriberClient::PubsubCommandBatch(
...
+  if (cluster_id.IsNil()) {
+    gcs_rpc_client_->GetClusterId(
+        rpc::GetClusterIdRequest(),
+        [this](const Status &status, const rpc::GetClusterIdReply &reply) {
+          auto cluster_id = ClusterID::FromBinary(reply.cluster_id());
```
Comment (jjyao) [inline src/ray/gcs/gcs_client/gcs_client.cc:103]: how do you handle failures?

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 24 above)
[the remark this answers is not in the document]
Comment (jjyao) [reply src/ray/gcs/gcs_client/gcs_client.cc:98]: I missed it: how does the GetClusterId call go through node_accessor_?

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 23 above)
[the remark this answers is not in the document]
Comment (jjyao) [reply src/ray/gcs/gcs_server/gcs_server.cc]: > The alternative to future is to use the sync API

I feel the code will be cleaner this way: we can just pass the actual cluster id around after the sync API?

File: src/ray/gcs/gcs_server/gcs_node_manager.h [hunk 28]
```diff
@@ -167,6 +173,8 @@ class GcsNodeManager : public rpc::NodeInfoHandler {
   std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage_;
   /// Raylet client pool.
   std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool_;
+  /// Cluster ID.
+  const ClusterID &cluster_id_;
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_node_manager.h:177]: I think it's safer to remove `&` to avoid potential dangling pointer. Copying ClusterID is cheap anyway.

File: src/ray/protobuf/gcs_service.proto [hunk 29]
```diff
@@ -177,6 +177,13 @@ service ActorInfoGcsService {
   rpc KillActorViaGcs(KillActorViaGcsRequest) returns (KillActorViaGcsReply);
 }
 
+message GetClusterIdRequest {}
+
+message GetClusterIdReply {
+  GcsStatus status = 1;
```
Comment (jjyao) [inline src/ray/protobuf/gcs_service.proto:183]: Do we need this? isn't it always OK?

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 24 above)
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/gcs/gcs_client/gcs_client.cc:98]: I think here, the `RegisterClient` is put into NodeManager RPC service. NodeInfoAccess calls that service.

You need some place for the call, and right now it's just to NodeInfoAccess.

An alternative way is to create a separate rpc service for that work only.

Or, we just put it into `GetSystemConfig`.

File: src/ray/gcs/gcs_client/test/gcs_client_test.cc [hunk 30]
```diff
@@ -987,9 +1003,21 @@ TEST_P(GcsClientTest, TestEvictExpiredDestroyedActors) {
   }
 }
 
+TEST_P(GcsClientTest, TestGcsAuth) {
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/test/gcs_client_test.cc:1006]: Can you test some negative tests? Like the auth is set incorrectly and GCS just reject the call?

File: src/ray/rpc/worker/core_worker_server.h [hunk 31]
```diff
@@ -41,10 +41,10 @@ namespace rpc {
   RPC_SERVICE_HANDLER_SERVER_METRICS_DISABLED(CoreWorkerService, PubsubCommandBatch, -1) \
   RPC_SERVICE_HANDLER_SERVER_METRICS_DISABLED(                                           \
       CoreWorkerService, UpdateObjectLocationBatch, -1)                                  \
-  RPC_SERVICE_HANDLER_SERVER_METRICS_DISABLED(                                           \
-      CoreWorkerService, GetObjectLocationsOwner, -1)                                    \
   RPC_SERVICE_HANDLER_SERVER_METRICS_DISABLED(                                           \
       CoreWorkerService, ReportGeneratorItemReturns, -1)                                 \
+  RPC_SERVICE_HANDLER_SERVER_METRICS_DISABLED(                                           \
```
Comment (iycheng) [inline src/ray/rpc/worker/core_worker_server.h:46]: nit: revert this change?

File: src/ray/rpc/client_call.h [hunk 32]
```diff
@@ -258,6 +274,12 @@ class ClientCallManager {
     return call;
   }
 
+  // For testing purposes only.
+  void StampContext(grpc::ClientContext &context) {
```
Comment (iycheng) [inline src/ray/rpc/client_call.h:278]: Try to move to test file.
We can have a getter function to retrieve cluster_id in GcsClient

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 27 above)
Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc:103]: check added

File: src/ray/gcs/gcs_client/test/gcs_client_test.cc (hunk 30 above)
Comment (vitsai) [reply src/ray/gcs/gcs_client/test/gcs_client_test.cc:1006]: That's tested in the next PR :)

File: src/ray/gcs/gcs_server/gcs_node_manager.h (hunk 28 above)
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_node_manager.h:177]: It won't be dangling because gcs_node_manager lives in gcs_server

File: src/ray/protobuf/gcs_service.proto (hunk 29 above)
Comment (vitsai) [reply src/ray/protobuf/gcs_service.proto:183]: I think having Status is convention, you can see other Replys in here that even have a "not populated" comment. Always useful for future evolution when Status will be necessary.

File: src/ray/raylet/main.cc [hunk 33]
```diff
@@ -170,153 +170,156 @@ int main(int argc, char *argv[]) {
   RAY_CHECK_OK(gcs_client->Connect(main_service));
   std::unique_ptr<ray::raylet::Raylet> raylet;
 
-  RAY_CHECK_OK(gcs_client->Nodes().AsyncGetInternalConfig(
-      [&](::ray::Status status,
-          const boost::optional<std::string> &stored_raylet_config) {
-        RAY_CHECK_OK(status);
-        RAY_CHECK(stored_raylet_config.has_value());
-        RayConfig::instance().initialize(stored_raylet_config.get());
+  auto f = std::async(std::launch::async, [&]() {
```
Comment (iycheng) [inline src/ray/raylet/main.cc:173]: Is this necesary? can you just call `main_service` at the end?

Btw, don't bind `boost::asio::io_service::work main_work(main_service);` at the beginning.

It's very similar to [this](https://github.com/ray-project/ray/blob/ec85f874d99c2c9c44415437c91c8370e0af7c5a/src/ray/gcs/gcs_server/gcs_server.cc#L70-L82)

File: src/ray/rpc/client_call.h (hunk 18 above)
Comment (vitsai) [reply src/ray/rpc/client_call.h:334]: Removed "generation"

File: src/ray/raylet/main.cc (hunk 33 above)
Comment (vitsai) [reply src/ray/raylet/main.cc:173]: `main_service` needs to be running in order to process the `GetClusterId` RPC callback. We need to have Cluster ID by this point already in order to send this RPC. 

Currently, with future:
- This will deadlock if we do not launch it in a separate thread, because it will block on the cluster ID future, which will never get started

Without future:
- This will still deadlock if we switch to a sync API because unless we start main_service, send the GetClusterID RPC, stop main_service, and continue.

**Comment by iycheng** [comment]
After reading the code around, I feel maybe we can put the token retrieving logic into `GetInternalConfig` because the current code base has already had this logic: before starting raylet, we retrieve the config; before starting GCS, we put the config.
If we store it there, what we need is:

- before starting GCS, we get and then if not there, put.
- in raylet, after retrieving the config, recreated a client with the token.

I feel in this way, we can avoid the future/promise completely because the token can be passed to GcsClient directly.

**Comment by vitsai** [comment]
Ok so @iycheng and I discussed offline several alternative plumbing approaches that would avoid using futures or otherwise be simpler+cleaner, including what he mentioned above. 

- Using the config to pass would be nice as it avoids an RPC, but can be done with or without future.
- We could also keep the same approach, but instead of passing a future, just populate later and fatal if not populated in the future.get() part of the code
- A promising approach could be to pass the token around in the services.py layer but would entail some more plumbing

In any case I would advocate for merging this soon and layering changes on top of that as it touches 35 files and isn't that small, especially if there's significant additional plumbing changes.
